blog/en-us/spring-boot-dubbo-start-stop-analysis.md

```diff
@@ -0,0 +1,223 @@
+- There are alse some threads about GC, and threads that start with `VM`, They are also non-daemon threads, but they are most likely the JVM's own threads, which are ignored here.
+
+In summary, we can infer that it is likely that the `container-0` and `container-1` cause the JVM to not exit. Now let's search through the source code to find out who created the two threads.
+
+By the source code analysis of Spring-boot, we can find these code on `org.springframework.boot.context.embedded.tomcat.TomcatEmbeddedServletContainer`'s `startDaemonAwaitThread`.
+
+```java
+    private void startDaemonAwaitThread() {
+        Thread awaitThread = new Thread("container-" + (containerCounter.get())) {
+
+            @Override
+            public void run() {
+                TomcatEmbeddedServletContainer.this.tomcat.getServer().await();
+            }
+
+        };
+        awaitThread.setContextClassLoader(getClass().getClassLoader());
+        awaitThread.setDaemon(false);
+        awaitThread.start();
+    }
+```
+
+Let's add a breakpoint in this method, and focus on the call stack:
+
+```plain
+initialize:115, TomcatEmbeddedServletContainer (org.springframework.boot.context.embedded.tomcat)
+<init>:84, TomcatEmbeddedServletContainer (org.springframework.boot.context.embedded.tomcat)
+getTomcatEmbeddedServletContainer:554, TomcatEmbeddedServletContainerFactory (org.springframework.boot.context.embedded.tomcat)
+getEmbeddedServletContainer:179, TomcatEmbeddedServletContainerFactory (org.springframework.boot.context.embedded.tomcat)
+createEmbeddedServletContainer:164, EmbeddedWebApplicationContext (org.springframework.boot.context.embedded)
+onRefresh:134, EmbeddedWebApplicationContext (org.springframework.boot.context.embedded)
+refresh:537, AbstractApplicationContext (org.springframework.context.support)
+refresh:122, EmbeddedWebApplicationContext (org.springframework.boot.context.embedded)
+refresh:693, SpringApplication (org.springframework.boot)
+refreshContext:360, SpringApplication (org.springframework.boot)
+run:303, SpringApplication (org.springframework.boot)
+run:1118, SpringApplication (org.springframework.boot)
+run:1107, SpringApplication (org.springframework.boot)
+main:35, DubboConsumerDemo (com.alibaba.boot.dubbo.demo.consumer.bootstrap)
+```
+
+It can be seen that during the startup process of the Spring-boot application, the above method is executed since the Tomcat exposes the HTTP service by default, and all threads started by Tomcat are daemon threads by default, such as the Acceptor of the listening request, working thread pool etc.The JVM will also exit after the startup is completed if there is no control here. Therefore, it is necessary to explicitly start a thread and continue to wait under certain conditions. We can avoid thread exit by this way.
+
+Let's dig deeper. In Tomcat's `this.tomcat.getServer().await()` method, how the thread doesn‘t exit. The irrelevant code is removed here for the convenience of reading.
```

Comment:
Let's dig deeper to find out how the thread stay alive in Tomcat's `this.tomcat.getServer().await()` method.